tracer/test/Datadog.Trace.Tests/Telemetry/Collectors/ProductsTelemetryCollectorTests.cs

```diff
@@ -0,0 +1,116 @@
+﻿// <copyright file="ProductsTelemetryCollectorTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Linq;
+using Datadog.Trace.Telemetry;
+using FluentAssertions;
+using Xunit;
+
+namespace Datadog.Trace.Tests.Telemetry;
+
+public class ProductsTelemetryCollectorTests
+{
+    [Fact]
+    public void DoesNotHaveChangesWhenNoCalls()
+    {
+        var collector = new ProductsTelemetryCollector();
+
+        collector.HasChanges().Should().BeFalse();
+        collector.GetData().Should().BeNull();
+    }
+
+    [Theory]
+    [MemberData(nameof(Data.AllEnabledPermutations), MemberType = typeof(Data))]
+    public void HasChangesWhenSingleProductChanged(bool? appsecEnabled, bool? profilerEnabled, bool? debuggerEnabled)
+    {
+        var collector = new ProductsTelemetryCollector();
+        if (appsecEnabled is not null)
+        {
+            collector.ProductChanged(TelemetryProductType.AppSec, enabled: appsecEnabled.Value, error: null);
+        }
+
+        if (profilerEnabled is not null)
+        {
+            collector.ProductChanged(TelemetryProductType.Profiler, enabled: profilerEnabled.Value, error: null);
+        }
+
+        if (debuggerEnabled is not null)
+        {
+            collector.ProductChanged(TelemetryProductType.DynamicInstrumentation, enabled: debuggerEnabled.Value, error: null);
+        }
+
+        var data = collector.GetData();
+        (data?.Appsec?.Enabled).Should().Be(appsecEnabled);
+        (data?.Profiler?.Enabled).Should().Be(profilerEnabled);
+        (data?.DynamicInstrumentation?.Enabled).Should().Be(debuggerEnabled);
+    }
+
+    [Fact]
+    public void DoesNotHaveChangesIfNoChanges()
+    {
+        // This behaviour isn't _necessary_ but it documents the current existing behaviour
+        var collector = new ProductsTelemetryCollector();
+
+        collector.ProductChanged(TelemetryProductType.AppSec, enabled: true, error: null);
+        collector.HasChanges().Should().BeTrue();
+        collector.GetData().Should().NotBeNull();
+
+        collector.HasChanges().Should().BeFalse();
+        collector.GetData().Should().BeNull();
+    }
+
+    [Fact]
+    public void HasChangesWhenSettingTheSameValue()
+    {
+        // This behaviour isn't _necessary_ but it documents the current existing behaviour
+        var collector = new ProductsTelemetryCollector();
+
+        collector.ProductChanged(TelemetryProductType.AppSec, enabled: true, error: null);
+        collector.HasChanges().Should().BeTrue();
+        collector.GetData().Should().NotBeNull();
+
+        collector.ProductChanged(TelemetryProductType.AppSec, enabled: true, error: null);
```

Comment:
Dumb question sorry but are you sure we should not do like integrations and they want us to report only when situation has changed? (even though the impact should be lower so wouldn't be a big deal to report several times...)

Reply:
This _is_ like integrations, and we _do_ only report when the situation has changed 🤔

Reply:
Wait what? I must be missing something. Here the test records the same status twice (appsec enabled without error) and both time HasChanges is true and we can collect data. 
So I assume we do report the same state twice, no?

Reply:
_oh_ sorry, if we call `ProductChanged` again, then yeah, you're right. I meant that we _only_ report the product that was changed, not all of them. Given that can't actually happen in a real app currently, I'm distinctly "meh" about worrying about it 😉